docs/proposals/devfile.md

```diff
@@ -0,0 +1,64 @@
+
+# Devfile support in odo
+
+To better integrate odo with other developer tooling we need to align on one common definition of a developer workspace and application lifecycle.
+
+[Eclipse Che](https://github.com/eclipse/che) project already created and started using [Devfile](https://github.com/redhat-developer/devfile). Devfile is a YAML file that defines developer workspace in Che environment. It is structured in such a way that can be used also in odo.
+
+Implementing Devfile support in odo will create a great benefit for users, as it will allow them to easily switch tools with no additional configuration. Adding support for new languages will be also much easier with devfiles, as it will be only a matter of using the right devfile template.
+
+## Implementation plan
+
+### Consuming current version of Devfile in odo
+
+1. Create a golang library for parsing and reading data from Devfile
+    * Devfile is versioned file so the design of a library needs to be able to handle and support multiple Devfile versions at the same time
+
+        **Parser high level design**
+        * The parser would read the provided devfile.
+        * From the devfile, parser would look for the API Version of the devfile and validate whether the API version in supported in odo.
+        * The devfile schema is based on <http://json-schema.org/draft-07/schema#>
+        * Hence, the parser would validate the integrity of the provided devfile with the devfile json schema of the respective API version.
+        * Furthermore, parser would try to map devfile sections (projects, components, commands etc) to their respective Golang structs.
+        * The idea is to have same Golang structs for the devfile sections across different API version contingent to the condition that the sections themselves are compatible across the API versions..
+        * This is very similar to the way Golang structs are handled across different GVK versions in kubernetes. [https://github.com/kubernetes/kubernetes/blob/ded2ff39c381b122951dd103dafb291755791921/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L32](https://github.com/kubernetes/kubernetes/blob/ded2ff39c381b122951dd103dafb291755791921/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L32)
+        * The implementation details for this with respect to odo are being worked upon and once confirmed would be updated here.
+
+2. Implement experimental support for reading current version of a Devfile and bootstrap odo components from a Devfile.
+    * If there is a devfile.yml and no .odo/config.yaml `odo push` will use Devfile.yaml and create necessary resources in Kubernetes cluster.
```

Comment:
What about the case when both `devfile.yml` and `.odo/config.yml` exist?

Reply:
There will be an experimental feature flag (https://github.com/openshift/odo/issues/2536) if enabled and both files exists, the devfile.yaml will have higher priority. If experimental is not enabled it will ignore the devfile.yaml file

Reply:
Ack, thanks.